src/Plugin/Commerce/PaymentGateway/ExpressCheckout.php

```diff
@@ -0,0 +1,570 @@
+    }
+
+    $flow = 'ec';
+    // Build a name-value pair array for this transaction.
+    $nvp_data = [
+      'METHOD' => 'SetExpressCheckout',
+
+      // Default the Express Checkout landing page to the Mark solution.
+      'SOLUTIONTYPE' => 'Mark',
+      'LANDINGPAGE' => 'Login',
+
+      // Disable entering notes in PayPal, as we don't have any way to accommodate
+      // them right now.
+      'ALLOWNOTE' => '0',
+
+      'PAYMENTREQUEST_0_PAYMENTACTION' => $payment_action,
+      'PAYMENTREQUEST_0_AMT' => $this->formatNumber($amount),
+      'PAYMENTREQUEST_0_CURRENCYCODE' => $payment->getAmount()->getCurrencyCode(),
+      'PAYMENTREQUEST_0_INVNUM' => $order->id() . '-' . REQUEST_TIME,
+
+      // Set the return and cancel URLs.
+      'RETURNURL' => $extra['return_url'],
+      'CANCELURL' => $extra['cancel_url'],
+    ];
+
+    $order_express_checkout_data = $order->getData('paypal_express_checkout');
+    if (!empty($order_express_checkout_data['token'])) {
+      $nvp_data['TOKEN'] = $order_express_checkout_data['token'];
+    }
+
+    // Get the order line items
+    $n = 0;
+    foreach ($order->getItems() as $item) {
+      $item_amount = $item->getUnitPrice()->getNumber();
+      $nvp_data += [
+        'L_PAYMENTREQUEST_0_NAME' . $n => $item->getTitle(),
+        'L_PAYMENTREQUEST_0_AMT' . $n => $this->formatNumber($item_amount),
+        'L_PAYMENTREQUEST_0_QTY' . $n => $item->getQuantity(),
+      ];
+      $n++;
+    }
+
+    // If reference transactions are enabled and a billing agreement is supplied...
+    if (!empty($configuration['reference_transactions']) &&
+      !empty($configuration['ba_desc'])) {
+      $nvp_data['BILLINGTYPE'] = 'MerchantInitiatedBillingSingleAgreement';
+      $nvp_data['L_BILLINGTYPE0'] = 'MerchantInitiatedBillingSingleAgreement';
+      $nvp_data['L_BILLINGAGREEMENTDESCRIPTION0'] = $configuration['ba_desc'];
+    }
+
+    // If Express Checkout Account Optional is enabled...
+    if ($configuration['solution_type'] != 'Mark') {
+      // Update the solution type and landing page parameters accordingly.
+      $nvp_data['SOLUTIONTYPE'] = 'Sole';
+
+      if ($configuration['solution_type'] == 'SoleBilling') {
+        $nvp_data['LANDINGPAGE'] = 'Billing';
+      }
+    }
+
+    // ToDo Shipping data.
+    $nvp_data['NOSHIPPING'] = '1';
+
+    // Overrides specific values for the BML payment method.
+    if ($flow == 'bml') {
+      $nvp['USERSELECTEDFUNDINGSOURCE'] = 'BML';
+      $nvp['SOLUTIONTYPE'] = 'SOLE';
+      $nvp['LANDINGPAGE'] = 'BILLING';
+    }
+
+    // Make the PayPal NVP API request.s
+    return $this->apiRequest($nvp_data);
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function apiGetExpressCheckoutDetails(OrderInterface $order) {
+    // Get the Express Checkout order token.
+    $order_express_checkout_data = $order->getData('paypal_express_checkout');
+    if (empty($order_express_checkout_data['token'])) {
+      return FALSE;
```

Comment:
i think this is not ok. i'll move into onReturn (and throw an exception) before the call of this.